src/v/cluster/types.cc

```diff
@@ -437,6 +437,27 @@ cluster::join_request adl<cluster::join_request>::from(iobuf_parser& in) {
     return cluster::join_request(adl<model::broker>().from(in));
 }
 
+void adl<cluster::join_node_request>::to(
+  iobuf& out, cluster::join_node_request&& r) {
+    adl<uint8_t>().to(out, r.current_version);
+    adl<int64_t>().to(out, r.logical_version);
+    adl<model::broker>().to(out, std::move(r.node));
+}
+
+cluster::join_node_request adl<cluster::join_node_request>::from(iobuf io) {
+    return reflection::from_iobuf<cluster::join_node_request>(std::move(io));
+}
+
+cluster::join_node_request
+adl<cluster::join_node_request>::from(iobuf_parser& in) {
+    auto version = adl<uint8_t>().from(in);
+    vassert(version >= 1, "Malformed join_node_request");
```

Comment:
is the inequality wrt `current_version` or `1`?

Reply:
With 1 -- this is validating the encoding version of the struct rather than any of the logical versions.  There's a note about the dual meaning of 'version' in the struct definition, but it's kind of unavoidably a bit confusing anyplace we're reasoning about multiple versions in one place.